File: C++/src/VdSQLiteXYZ.cpp
/*!
 *  \brief		class for writing data to sqlite database
 *
 *  \author		Florian Timm
 *  \version	2020.11.28
 *  \copyright	MIT License
 */

#include "VdSQLiteXYZ.h"
#include <stdio.h>
#include <iostream>
#include <sqlite3.h>
using namespace std;

VdSQLiteXYZ::VdSQLiteXYZ(string filename) {
	/**
	 * constructor
	 * @param filename name of db file
	 */
	this->open(filename);
	sqlite3 *db;
}

void VdSQLiteXYZ::open(string filename) {
	/**
	 * opens a new db file
	 * @param filename name of db file
	 */
	char *zErrMsg = 0;
	filename = this->makeFilename(string("db"), filename);
	sqlite3_open(filename.c_str(), &this->db);

	string sql =
			"CREATE TABLE IF NOT EXISTS xyz_data (id INTEGER PRIMARY KEY AUTOINCREMENT, time FLOAT, x FLOAT, y FLOAT, z FLOAT)";
	sqlite3_exec(this->db, sql.c_str(), 0, 0, &zErrMsg);
	sqlite3_prepare_v2(db,
			"INSERT INTO xyz_data (time, x, y, z) VALUES (?, ?, ?, ?)",
			-1, &stmt, 0);
}

void VdSQLiteXYZ::write() {
	/** writes data to database */
	char *zErrMsg = 0;
	sqlite3_exec(this->db, "BEGIN TRANSACTION;", 0, 0, &zErrMsg);
	for (VdPoint p : this->writingQueue) {
		sqlite3_bind_double(stmt, 1, p.getTime());
		VdXYZ xyz = p.getXYZ();
		sqlite3_bind_double(stmt, 2, xyz.getX());
		sqlite3_bind_double(stmt, 3, xyz.getY());
		sqlite3_bind_double(stmt, 4, xyz.getZ());
		sqlite3_step(stmt);
		sqlite3_reset(stmt);
	}
	sqlite3_exec(this->db, "END TRANSACTION;", 0, 0, &zErrMsg);
	this->clearWritingQueue();
}
void VdSQLiteXYZ::close() {
	/** close database */
	sqlite3_close(db);
}

File: Python/vdSQLiteXYZ.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
@author: Florian Timm
@version: 2020.11.28
"""

import sqlite3
from vdFile import VdFile
from vdPoint import VdPoint


class VdSQLiteXYZ(VdFile):

    """ class for writing data to sqlite database """

    def _open(self, filename):
        """
        opens a new db file
        :param filename: name of db file
        :type filename: str
        """
        filename = self._make_filename("db", filename)
        print(filename)
        self.__db = sqlite3.connect(filename)
        self.__cursor = self.__db.cursor()
        self.__cursor.execute("CREATE TABLE IF NOT EXISTS raw_data ("
                              "id INTEGER PRIMARY KEY AUTOINCREMENT,"
                              "time FLOAT,"
                              "x FLOAT,"
                              "y FLOAT,"
                              "z FLOAT)")
        self.__db.commit()

    def write(self):
        """ writes data to database """
        insert = []
        for p in self.writing_queue:
            x, y, z = p.get_xyz()
            insert.append((p.time, x, y, z))

        self.__cursor.executemany("INSERT INTO raw_data ("
                                  "time, x, y, z) "
                                  "VALUES (?, ?, ?, ?)", insert)
        self.__db.commit()
        self.clear_writing_queue()

    def close(self):
        """ close database """
        self.__db.close()
